### ISSUE TITLE
Agents going offline (in general)

### ISSUE THREAD

**Issue by southeasterntech** [Ylianst/MeshCentral#2831]
Guys, 

There's been something that's been bugging me for some time and I'm not sure how to communicate it exactly.  I also don't have any reproducible evidence of it.

In general agents don't seem to stay reliably connected.  

Much like the Windows 7\2008 Agent disconnect issue.... which we still (embarrassingly) have a number of servers out there behind firewalls that are still not updated... I'd love to see the agents stay connected on this, but I get it.  EOL.

Today I saw a linux machine of ours...that is a tinkerboard, and runs a dashboard in our office.  It's had mesh since 2 years ago, and has been there reliably, until I noticed this week, it was no longer connected.  This machine hasn't changed or even updated in any way in 2 years.  Why does this happen? I get the odd MC service hanging on the odd machine... but this seems more frequent than that.

I reinstalled the above tinkerboard agent with the latest one... and it pops up as a second entry.  I always have to go back and delete the old device... (See screen)

Secondly, I see a really large amount of disconnected devices in MC (see screen).  I'd think that, well..those are just off.... hard to keep track of so many computers.  I ran a software deployment tool on a customer's DC to push MC out to the entire domain.  Something like 10 computers lit up in blue again.  So they were there all along, just disconnected in MC.  I had to reinstall the agent to get them to show up again.    I know there can't be that many computers that are just off, so it makes it hard to know what's happening and frustrating to open up another remote tool to just restart the mesh service\upgrade the agent.

Thirdly.  There are a good number of servers that we manage.  Many of them regularly show disconnected in MC, and they're 2016.  It requires restarting the agent service at minimum or reinstall to get them reconnected.

The bottom line is that I don't feel the agent connection status is reliable.  Maybe there's a good reason for this, such as, "we had a problem with older versions of MC and the new agent is better."

1.  Is there a way to make the agent more robust?  Some way to check the service status at an interval and restart it automatically?  I think there's a self update feature in the agent, but I'm not convinced that's working as reliably as it should.  

2.  Is there a way to cause a previously connected agent to show up again as the same agent, so we don't have to find the old disconnected agent and delete it?

3.  It would be great if there was a deployment tool that could push out the agents again... yes I know I can use AD\GP and a host of other deployment options... it'd be nice if there was some built in feature... 

I realize it's probably a very challenging subject and you can't fix a problem you can't see.   I just feel like the agent reliability needs to be improved.   I'd love for one of you super smart devs to tell me, here's where you have it configured wrong.  Anyway, thanks for listening.

While the above may sound ungrateful, since this tool is currently free and BR and YSH spend tons of time improving things already..... I really don't mean to be.  Just wanted to start a conversation on this subject.

(BTW, I just noticed that 2 of my 2016 servers that we monitor for online availability, are not active in mesh.  Just reinstalled and they're back again)

![mesh-tinker](https://user-images.githubusercontent.com/45638572/123964126-22a24300-d981-11eb-9c10-20b272ca8e95.JPG)
![mesh-tinker2](https://user-images.githubusercontent.com/45638572/123964128-233ad980-d981-11eb-9ae6-fac6ab60c7c7.JPG)

Shane-Southeastern Tech

**Comment by nzalev** [comment]
This is an issue that I have come across as well, including having to re-install and remove the previous entry for the agent.

**Comment by southeasterntech** [comment]
In the interest of actually taking this conversation somewhere tangible.  I have a few cases to present.  Standby while I compile

**Comment by southeasterntech** [comment]
CASE 1. 
-(Flex-3)  WIndows 8.1 x64
-Showing Offline in MC
-Showing Online in another RMM
-Mesh Service is running
-Can Ping Server
![flex-3-service-ping](https://user-images.githubusercontent.com/45638572/124137700-eee41d80-da53-11eb-8110-f43fafb576de.png)
![flex-3-mesh](https://user-images.githubusercontent.com/45638572/124137703-eee41d80-da53-11eb-8cf1-902f4a3d5f4e.JPG)

**Comment by southeasterntech** [comment]
CASE 2.
-(XCP-NG 8.0.0) Linux Distro Based on Centos (Forked from Citrix Xenserver)
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server
![xcp-ng-service-running](https://user-images.githubusercontent.com/45638572/124140204-471c1f00-da56-11eb-9922-597d84adbb92.JPG)
![xcp-ng 8 0 0 - ping](https://user-images.githubusercontent.com/45638572/124140206-47b4b580-da56-11eb-882f-119778a32f1c.png)

**Comment by southeasterntech** [comment]
CASE .
-(XCP-NG 8.0.0) Linux Distro Based on Centos (Forked from Citrix Xenserver)
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server
-Shows multiple connections with each reinstall
![xcp-ng-disconnected-multiple](https://user-images.githubusercontent.com/45638572/124140559-94988c00-da56-11eb-9d3a-7de0c67fe984.JPG)

**Comment by southeasterntech** [comment]
CASE 3.
-(XOA) Debian 9
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server
![xoa-agent-ping](https://user-images.githubusercontent.com/45638572/124141707-9dd62880-da57-11eb-878b-cfe426e5ecab.png)
![xoa-agent-running](https://user-images.githubusercontent.com/45638572/124141709-9e6ebf00-da57-11eb-921d-22ba65919fa9.png)
![xoa-offline](https://user-images.githubusercontent.com/45638572/124141710-9e6ebf00-da57-11eb-9022-447cbe76e3ad.JPG)

**Comment by southeasterntech** [comment]
CASE 4.
-(Marketing Dashboard-Internal LAN) Armbian on Tinkerboard (Ubuntu 18.04.1 LTS)
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server
![arbian-tinkerboard-marketing-ping](https://user-images.githubusercontent.com/45638572/124146562-e263c300-da5b-11eb-9bcc-b002845a4071.png)
![arbian-tinkerboard-marketing-running](https://user-images.githubusercontent.com/45638572/124146569-e2fc5980-da5b-11eb-9c53-8a1ddfccd490.JPG)
![arbian-tinkerboard-marketing-stats](https://user-images.githubusercontent.com/45638572/124146570-e2fc5980-da5b-11eb-87db-dee1069ba0a8.JPG)

**Comment by southeasterntech** [comment]
Config.JSON for reference
![config json](https://user-images.githubusercontent.com/45638572/124147100-64ec8280-da5c-11eb-94ec-0b94158c5b75.png)

**Comment by southeasterntech** [comment]
That's probably plenty for now though I could prob dig up other examples on 2016\Win10

**Comment by southeasterntech** [comment]
@Ylianst just wanted to make sure I updated this ticket with something tangible... Thanks as Always.  You guys rock

**Comment by krayon007** [comment]
One thing you can do, is to navigate to the install path of the agent, and run the command:
`./meshagent -state`

This will connect to the Agent's IPC, and display some diagnostic information. If the agent is responsive, it should show something useful, including what the agent think's its connection state is.

**Comment by southeasterntech** [comment]
Thanks Bryan,

CASE 1.
-(Flex-3) WIndows 8.1 x64
-Showing Offline in MC
-Showing Online in another RMM
-Mesh Service is running
-Can Ping Server

![image](https://user-images.githubusercontent.com/45638572/124952363-2c135700-dfe2-11eb-9512-4e7e9466fa38.png)

**Comment by southeasterntech** [comment]
CASE 2.
-(XCP-NG 8.0.0) Linux Distro Based on Centos (Forked from Citrix Xenserver)
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server
-Shows multiple previous connections

![image](https://user-images.githubusercontent.com/45638572/124952898-9f1ccd80-dfe2-11eb-80e2-7ea6a715ae1f.png)

**Comment by southeasterntech** [comment]
CASE 3.
-(XOA) Debian 9
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server
![image](https://user-images.githubusercontent.com/45638572/124953801-73e6ae00-dfe3-11eb-8c58-489a8797da89.png)

**Comment by southeasterntech** [comment]
CASE 4.
-(Marketing Dashboard-Internal LAN) Armbian on Tinkerboard (Ubuntu 18.04.1 LTS)
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server

![image](https://user-images.githubusercontent.com/45638572/124954672-549c5080-dfe4-11eb-9aa6-7f401191ac18.png)

**Comment by southeasterntech** [comment]
Bryan, so it seems in all these cases the agent can't be contacted.  Surely there is something in common here that points to a problem that could be solved to improve reliability of the agent?
Thanks
Shane

**Comment by krayon007** [comment]
can you use `gcore` to create a dump of the agent's pid? That would be very helpful in trying to track down what is going on.

**Comment by southeasterntech** [comment]
Can you provide a little guidance on how\where to use gcore?  Can't seem to find any info on it.

**Comment by southeasterntech** [comment]
disregard, thought it was a mesh thing.... only for linux right?

**Comment by southeasterntech** [comment]
CASE 3.
-(XOA) Debian 9
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server

so might need a little assistance with this.  
I installed GDB, determined the PID which was 64221...  

then tried to run gcore:
 gcore -o /tmp/mesh 64221 


sudo gcore -o /tmp/mesh 64221
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
__lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
95      ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or directory.
Saved corefile /tmp/mesh.64221


I have that file attached, but not sure I did it right.  Please let me know then I can do it on the others

**Comment by nzalev** [comment]
I'm happy to see some movement on this. It's been a pretty long standing issue from what I've seen.

**Comment by LPJon** [comment]
I have a question for @southeasterntech . So are all this machines directly connected to MeshCentral Server, as in they are on the same LAN? Or are they across an internet connection? My thoughts on this are that I believe MeshCentral and Meshagent listen to each other. If MeshCentral cant hear Meshagent then MeshCentral would assume the connection died and show disconnected. Did you ever check if the MeshCentral Server could ping the Agents? This looks like a possible firewall or DPI (Deep Packet Inspection) issue. It could even be a webserver issue with too many connections. Are any of these using reverse proxies? If so....which ones? I currently have about 150 systems and they are working just fine. They range from Windows, Windows Server, Debian 9 & 10, ubuntu, Zorin, CentOS 7, 8 & 8.2, XCP-ng and a lot more. I just want to help with this too. There just isn't enough information to even know where to start looking. Mostly, this looks like it may be a networking issue where a connection on one side is kept open while the connection on the other side is closed. My network engineering experience gut instinct tells me that's a firewall somewhere miss-behaving or a security tool running on one at the very least but reverse proxies can cause this too.

**Comment by southeasterntech** [comment]
@LPJon in All CASES above, the only device on the LAN is CASE4,

CASE 4.
-(Marketing Dashboard-Internal LAN) Armbian on Tinkerboard (Ubuntu 18.04.1 LTS)
-Showing Offline in MC
-Mesh Service is running
-Can Ping Server

CASES 1-3 are all external to the LAN, contacting the mesh server on the WAN interface.

I have not tried to see if the MC server can ping those agents, however, do note that in all CASES above, those remote LAN's all have other machines that are running meshagents and those are not having issues.  We do have 22 some Xenservers we monitor (Centos) and they always stay connected for awhile but then tend to drop off fairly reliably.

I wouldn't think it was DPI as I mentioned other machines on those remote LAN's aren't having issues.   But I can't put my finger on what the commonality is.  It seems random on all flavors of OS.

Now I am running caddy reverse proxy.... but running MongoDB

I really want to get to the bottom of this as well, as it never quite seems to go away.  I'd be happy to do some collaborative work, share screens, or whatever someone smarter than me needs to figure out why this happens.  

Of the 4 CASES I've presented above, I still haven't touched them (reinstalled agents, etc), so we can hopefully use that to figure it out.

Thanks to all.  @LPJon, feel free to ask anything else.

**Comment by southeasterntech** [comment]
@Ylianst @krayon007 
OK, still seeing this... today I'll present CASE5

wp.setechnical.net
Ubuntu 16.04.  Same as above..
LAN Computer
Mesh Service is running.
Can Ping server.   (Caddy Reverse Proxy) then redirected to Mesh Server.
Showing offline in Mesh since 7\6\2021
Restarting service has no effect

![image](https://user-images.githubusercontent.com/45638572/126202729-dc109926-f327-47a9-a7ff-bb06fe9d3784.png)

Reinstalled Mesh Agent with 
(wget "https://mesh.southeastern.biz/meshagents?script=1" --no-check-certificate -O ./meshinstall.sh || wget "https://mesh.southeastern.biz/meshagents?script=1" --no-proxy --no-check-certificate -O ./meshinstall.sh) && chmod 755 ./meshinstall.sh && sudo -E ./meshinstall.sh https://mesh.southeastern.biz 'ikY9q3aMAZP0pqzLwkB5jZnYZQU67ocvBBKfzC2XV@Mpa0W9wcvbbCuI44niEdSD' || ./meshinstall.sh https://mesh.southeastern.biz 'ikY9q3aMAZP0pqzLwkB5jZnYZQU67ocvBBKfzC2XV@Mpa0W9wcvbbCuI44niEdSD'

Now computer appears fine again in Mesh.....  so it's clearly not a networking or Reverse Proxy issue.
![image](https://user-images.githubusercontent.com/45638572/126203004-fe072182-7387-4044-be46-412b2d9bc65d.png)

**Comment by southeasterntech** [comment]
Guessing this isn't high priority for devs or too hard to figure out?  Question, is this my issue only or do other people have this issue too?  For me it's making mesh difficult to use.  I see this problem every day now that my attention is on it.

**Comment by southeasterntech** [comment]
CASE6
Server hosted at Liquid Web (WAN)
Windows 2016 Server
Mesh Agent Running, used to be there last week.  Not there now.
-Restarting Service has no effect
-Restarting Server has no effect.
-Reinstalling Mesh resolves
![image](https://user-images.githubusercontent.com/45638572/127173254-833da647-5d75-4803-ad92-14af03c7d162.png)
![image](https://user-images.githubusercontent.com/45638572/127173505-d7f4315b-95b1-41bb-9e5d-477149363765.png)
![image](https://user-images.githubusercontent.com/45638572/127175763-2932107c-dcf6-429f-a4b3-a0a62ce21d45.png)
![image](https://user-images.githubusercontent.com/45638572/127175897-743120d1-81a3-49b9-9023-2c0c16fc467d.png)

**Comment by nzalev** [comment]
I am mostly familiar with the server side of things, but I just downloaded the memory dump you uploaded. It immediately stands out: it's 300 MB. I did gcore dumps of agents and they are typically around 20 MB.

**Comment by southeasterntech** [comment]
@nzalev  It's entirely possible I did something wrong with gcore.  Thanks for looking at this BTW

============
so might need a little assistance with this.
I installed GDB, determined the PID which was 64221...

then tried to run gcore:
gcore -o /tmp/mesh 64221

sudo gcore -o /tmp/mesh 64221
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
__lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
95 ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or directory.
Saved corefile /tmp/mesh.64221
=====================

Can you provide any information on running gcore correctly?

**Comment by joeydc** [comment]
> Guessing this isn't high priority for devs or too hard to figure out? Question, is this my issue only or do other people have this issue too? For me it's making mesh difficult to use. I see this problem every day now that my attention is on it.

You're not alone. I've been dealing with this same issue for a very long time on 2 different MC installation. One at home behind a reverse proxy, with ~30 devices., and one at work with 50-60 devices hosted on a VPS behind no proxy. What a wonderful project, I just hope one day this issue will be fixed :)

**Comment by AntonAndreevichMoroz** [comment]
We also notice this sometimes on Windows machines. I cannot say that it is massive in our country, but in general there is such a situation.

**Comment by southeasterntech** [comment]
OK, good to know it's not just me

**Comment by GoodComputerGuy** [comment]
I also see this frequently. No rhyme or reason but invariably there are some Windows machines that I manage that just are not connected in MeshCentral any more. I have to restart the Mesh Agent service to get them to reconnect.

**Comment by southeasterntech** [comment]
...more dead horse beating... Gratefully :)

I re-ran the a deployment tool to re-deploy mesh to an active directory network.  and 3 computers had simply stopped connecting, and simply reappeared after reinstalling mesh.  See screens.  

Thanks

![deploytool](https://user-images.githubusercontent.com/45638572/128768025-05dab091-c240-41b0-be6e-a1a007f50cc1.png)
![Client Before Mesh Deploy](https://user-images.githubusercontent.com/45638572/128768026-f30b9d23-ff98-448e-89ce-a70edd9d0ed6.JPG)
![Client After Mesh Deploy](https://user-images.githubusercontent.com/45638572/128768028-3c08233c-5ecc-4060-9d60-ff26d19cc01f.JPG)

**Comment by southeasterntech** [comment]
> > Guessing this isn't high priority for devs or too hard to figure out? Question, is this my issue only or do other people have this issue too? For me it's making mesh difficult to use. I see this problem every day now that my attention is on it.
> 
> You're not alone. I've been dealing with this same issue for a very long time on 2 different MC installation. One at home behind a reverse proxy, with ~30 devices., and one at work with 50-60 devices hosted on a VPS behind no proxy. What a wonderful project, I just hope one day this issue will be fixed :)

@joeydc
did this situation improve for you after getting rid of the reverse proxy?  Wonder if that is complicating things?

**Comment by joeydc** [comment]
> > > Guessing this isn't high priority for devs or too hard to figure out? Question, is this my issue only or do other people have this issue too? For me it's making mesh difficult to use. I see this problem every day now that my attention is on it.
> > 
> > 
> > You're not alone. I've been dealing with this same issue for a very long time on 2 different MC installation. One at home behind a reverse proxy, with ~30 devices., and one at work with 50-60 devices hosted on a VPS behind no proxy. What a wonderful project, I just hope one day this issue will be fixed :)
> 
> @joeydc
> did this situation improve for you after getting rid of the reverse proxy? Wonder if that is complicating things?

@southeasterntech 
I don't believe it changed anything.
It happen way more often on the one without any proxy, but at the same time, it's also the one getting used the most.
It has to be something else.

**Comment by southeasterntech** [comment]
@Ylianst @krayon007 Guys, do you think the new agent update that Bryan just rolled out will help the above issues?  Love the new keyboard lock!
-Shane

**Comment by lokasoft** [comment]
We have installed MeshCentral about a week ago and use it to manage mainly windows servers on our remote sites. MeshCentral runs on Ubuntu 20 and behind HaProxy. Remote sites connect over the internet and remote servers are behind firewalls.
At the moment we manage around 60 servers. The remote sites are in part of the world where internet is less reliable.
So far we have seen two times where an agent was disconnected. As we also use another remote management tool we were able to remotely restart the MeshAgent service to let it reconnect.
Problem seems to be in the reconnection logic in case of network errors. The second time I had to restart the agent, it would not stop, I had to use taskmanager to kill the process after which it restarted and reconnected.
Is there anything we can check or can make reconnection more reliable?

**Comment by baishi** [comment]
I'm having exactly the same problem.

I only managed to install the service on 3 computers and 2 out of 3 reliably disconnects me every other 5-10 minutes.

The guest OS is Windows 10 Pro 21H1. I see other people have access to Mesh Agent 0.5.55 or 0.5.7x. After I download from my server, the agent version is 0.2.13.

The symptom is almost exactly the same as people above mentioned. Once the web client is left unattended for 3-5 minutes, there is a relatively high chance but seemingly random timeout that the browser client will show black screen with status still in "connected". If I disconnect and connect again, on a very high chance I can connect without "1 watching". If it happens again the watching will increase to 2 and so on.

After a while, if the watching reaches 3, especially reaches 5, the next connect will very likely stuck and then the Agent will show offline.

After managed to login to the server using other remote desktop software, the services.msc shows the Mesh Agent is running, while MeshAgent -state will show Unable to Contact Mesh Agent. I noticed that there are n MeshAgent.exe (n corresponding to the number of "watching" sessions) with zero CPU usage but the MeshAgent.exe related to the Service process uses roughly 26% on a very powerful Ryzen 3 processor. So I assume the Agent deadlock to some extent and could not respond to the server.

I tried to set agentPing: 120 without any success, however after I added agentPong: 30, I haven't seen any black screen since (for ~ 2 hours, but I'm quite busy working on the servers, so hard to tell if it solved my problem totally).

I saw #556 seemingly having the same symptom. Seemed it is a long standing problem.

**Comment by baishi** [comment]
OK I spoke too fast, the symptom relapsed. It seemed a NAT timeout issue, if you use it very heavily, it will stay online, but once left idle for too long, it will break. Sadly neither agentPing or agentPong solved it effectively.

I noticed something which I don't know if it is related. When I open a Terminal, the session will break every 2-3 minutes and I saw this in the console:
uncaughtException1: Error:  => EventEmitter.emit(): Event dispatch for 'end' on 'http.WebSocketStream.decoded' threw an exception: TypeError: invalid base value in method 'onTunnelClosed()'

**Comment by southeasterntech** [comment]
@baishi this is definitely what I'm seeing as well.  Especially the agent hang upon stop.  I wrote a batch script to nuke the service and then a reinstall gets it back up without a reboot.  But the reliability is very frustrating.  I'm not technical enough on the back end to troubleshoot it further.  I think we have 500+ machines in our system and half of them are disconnected, but with services running still.

**Comment by baishi** [comment]
> @baishi this is definitely what I'm seeing as well. Especially the agent hang upon stop. I wrote a batch script to nuke the service and then a reinstall gets it back up without a reboot. But the reliability is very frustrating. I'm not technical enough on the back end to troubleshoot it further. I think we have 500+ machines in our system and half of them are disconnected, but with services running still.

@southeasterntech would you help me quickly confirm if you are using a nginx reverse proxy? If so I found a solution at least for me case. Your mileage may vary.

**Comment by southeasterntech** [comment]
I'm using Caddy reverse proxy... could your solution work there?







Shane D. Lewis
Technical Director

Website: www.setechnical.net
Email: ***@***.***
Phone: (678)807-6156 x 701














From:   "Shi Bai" ***@***.***>
To:     "Ylianst/MeshCentral" ***@***.***>
Cc:     "southeasterntech" ***@***.***>, "Mention" 
***@***.***>
Date:   09/24/2021 11:48 AM
Subject:        Re: [Ylianst/MeshCentral] Agents going offline (in 
general) (#2831)




@baishi this is definitely what I'm seeing as well. Especially the agent 
hang upon stop. I wrote a batch script to nuke the service and then a 
reinstall gets it back up without a reboot. But the reliability is very 
frustrating. I'm not technical enough on the back end to troubleshoot it 
further. I think we have 500+ machines in our system and half of them are 
disconnected, but with services running still.

@southeasterntech would you help me quickly confirm if you are using a 
nginx reverse proxy? If so I found a solution at least for me case. Your 
mileage may vary.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by GoodComputerGuy** [comment]
I'm seeing the same issue with clients disconnecting still. New Mesh server
on AWS and I still have clients disconnected. I have to go in another way
and stop/start the Mesh service to get them to connect again.

**Comment by baishi** [comment]
OK I basically confirm this is VERY LIKELY to be an NAT mapping issue. For some (busy or memory limited) routers, the NAT table will be reused after some time (I guess is either 30 seconds or 60 seconds timer). So it appears to be extremely random and extremely dependent on usage/traffic. I have one very lousy server which is about 15 years old and never experience any problem. While I have some very decent servers behind a lousy router which constantly kick my a$$.

OK, now I got 15 minutes uptime without even one single instance of cut off.

The solution is to use agentPong: 30 in server config to ensure there is at least one packet going through every 30 seconds. The reason I chose 30 seconds is that my server's clock actually updates every minute; however, if there is no other traffic, it will crash every randomly. So 60 is definitely not enough.

Now the interesting part, agentPong WILL NOT WORK behind NGINX REVERSE PROXY. Other proxy software's mileage may vary.

The problem lies with the ssl buffer. I have done packet capture and saw the meshcentral server was actually quite chatty, but the nginx didn't emit many packets at all, at least not in timely fashion. I tried to increase proxy_read_timeout and proxy_send_timeout to 330s according to the document. I have a gut feel that the situation improved, but not realiable.

Only after I set ssl_buffer_size 8; to nginx, the nginx now reliably emits the agentPong packet and I haven't experienced any drop outs. The reason I chose 8 bytes is that the shortest message I spotted emitted by MeshCentral server is {action:"pong"}, 15 bytes. I tried to set the number to something greater than 1 to at least keep some performance gain.

@GoodComputerGuy If you are running nginx as reverse proxy, I believe ssl_buffer_size will help you if you haven't set agentPong yet.

@southeasterntech I saw you are using caddy, the configuration is definitely different. I'm trying to study and see if I can offer some help. In the meanwhile, I noticed in your message earlier, MeshCentral without reverse proxy also drops out. Can you confirm you have agentPong set on those servers?

The combination of agentPong and reliable emitting of those pong packets is the key to keep connection.

Because, at least for me, this has to happen over a lousy router, @Ylianst may never be able to reproduce this problem.

**Comment by baishi** [comment]
@southeasterntech I'm fairly confident after you set agentPong on your MeshCentral server and adjust flush_interval in you caddy proxy to something 30 seconds-ish. Your problem will be solved. Now I have 35 minutes idling uptime! It has never happened before!

**Comment by baishi** [comment]
@southeasterntech also make sure to use the Aug 10 version of the Mesh Agent. Some earlier version of the Agent has a bug that a agentPing/Pong request may kill the connection even faster.

**Comment by southeasterntech** [comment]
interesting post with similar issues
https://github.com/Ylianst/MeshCentral/issues/2779

**Comment by southeasterntech** [comment]
per this other post... adding the following to my config
    "browserPong": 30,
    "agentPong": 30,
    "agentPing":30,

which per @Ylianst is for helping with agent disconnection issues.

**Comment by baishi** [comment]
> interesting post with similar issues
> #2779

This is a totally different issue. Agent not reporting offline after losing network connectivity.

The truth is that the agent doesn't get a chance to report offline because the communication channel is cut, thus some idling timeout is important to report them MIA.

They are totally different issue than the one reported here.

I saw your post stating that even under LAN configuration the agent will drop off, which is quite weird and different from my case. However, agentPong is designed "almost exclusively" to solve this type of problem. Have a try and make sure either not using reverse proxy or Nginx with low ssl_buffer_size or caddy with flush_interval lower than 30 seconds. It really should help.

**Comment by baishi** [comment]
@southeasterntech is you problem any better by now?

I realised that by reading the source code that if agentPing is set then agentPong has no effect. Thus make sure setting agentPing with a valid value (in my case no larger than 30) should solve the problem quite effectively.

My observation of nginx not pushing the packet might be misleading since my Pong message is overridden by 120s' ping setup. May not be the nginx holding the message in buffer.

**Comment by southeasterntech** [comment]
I added the ping\pong as suggested and we are watching to see if it seems 
more stable.  Only been a couple of days.  I feel like it might be better? 
 Standby on this... will let you know shortly







Shane D. Lewis
Technical Director

Website: www.setechnical.net
Email: ***@***.***
Phone: (678)807-6156 x 701














From:   "Shi Bai" ***@***.***>
To:     "Ylianst/MeshCentral" ***@***.***>
Cc:     "southeasterntech" ***@***.***>, "Mention" 
***@***.***>
Date:   09/29/2021 09:51 AM
Subject:        Re: [Ylianst/MeshCentral] Agents going offline (in 
general) (#2831)




@southeasterntech is you problem any better by now?

I realised that by reading the source code that if agentPing is set then 
agentPong has no effect. Thus make sure setting agentPing with a valid 
value (in my case no larger than 30) should solve the problem quite 
effectively.

My observation of nginx not pushing the packet might be misleading since 
my Pong message is overridden by 120s' ping setup. May not be the nginx 
holding the message in buffer.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by silversword411** [comment]
I think I'm seeing the same issues listed here. When looking at the chrome, dev tools I see the websocket never loads control.ashx 

Enabling trace on the server, I get `id=Unknown` when restarting the mesh agent service (to try and get the agent to connect to the server and show as online - because what this is an issue the agent is offline).

![2022-01-31_221257 - mesh tracing](https://user-images.githubusercontent.com/13395348/151908698-47f18579-0986-424a-bab6-d3f3d91895ba.png)

I find restarting the meshcentral server will fix the issue...but that is just a workaround.

I think it's something to do with whatever causes "duplicateagentcount" to increment, but I haven't found a good way to debug who is causing those.

![2022-01-31_222421 - agentstats](https://user-images.githubusercontent.com/13395348/151909006-a3fb250f-95c8-418f-9101-49a601df795e.png)

Any detailed server or agent debug writeups are appreciated.

**Comment by bravery-only** [comment]
> CASE 1. -(Flex-3) WIndows 8.1 x64 -Showing Offline in MC -Showing Online in another RMM -Mesh Service is running -Can Ping Server ![flex-3-service-ping](https://user-images.githubusercontent.com/45638572/124137700-eee41d80-da53-11eb-8110-f43fafb576de.png) ![flex-3-mesh](https://user-images.githubusercontent.com/45638572/124137703-eee41d80-da53-11eb-8cf1-902f4a3d5f4e.JPG) ![flex-core](https://user-images.githubusercontent.com/45638572/124157326-a5e99480-da66-11eb-860e-2262a2f25667.JPG)



Add these three lines of code to the settings
      "browserpong": 30, 
      "agentpong": 15, 
      "agentping": 15, 

One question: I could not find this Flex-3 program that you showed in the photo? How can I get this program?

**Comment by southeasterntech** [comment]
@bravery-only Thanks for the comment.  I've added these items to the JSON some time ago, but the issue doesn't resolve.  

In the case of Flex-3, this is not a program.... this is the name of the computer I'm connected to.  
Regards

**Comment by bravery-only** [comment]
What is the program of these icons?
![Screenshot_8](https://user-images.githubusercontent.com/42680639/196601342-20fe5a0d-a4b4-4c9e-8c02-f2940950e1af.jpg)

**Comment by GoodComputerGuy** [comment]
That is NetSupport Manager.

**Comment by cfoellmann** [comment]
the settings change suggested by @southeasterntech / @Ylianst 
> "browserPong": 30,
> "agentPong": 30,
> "agentPing":30,

did "solve" this for us.
We are running the docker version of meshcentral + HAproxy as the reverse proxy. No "special" settings applied in HAproxy

**Comment by silversword411** [comment]
> did \"solve\" this for us.

I don't think you had the same problem with `id=unknown` there's another issue where if you have network equipment (routers, switches, VPNs etc) with short TCP session timers ping pong would fix that. This is different.

**Comment by si458** [comment]
closing as stale,
please try the latest version 1.1.21 and use node 18 or above,
if issues persist, please reply back

**Comment by pierrejoye** [comment]
little note for those coming after a search wondering why agents fail to connect. 

It just happened and it took me a few moments to remember the cert got renewed.

This will cause the agent to fail to connect (or the server to not valid the connection).

to avoid it, in general,  make sure CertUrl is set correctly, like if meshcentral is behind a proxy (the proxy needs to be used).
